### PR TITLE
Query.AI - Platform Integration

### PR DESCRIPTION
**Pull request by shaswat-indian**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
None

## Description
This is the initial pull request for adding Query.AI's (https://query.ai) platform integration to Demisto.

## Minimum version of Demisto
- [ ] 5.0.0
- [x] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation 

## Demisto Partner?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

**Comment by ShahafBenYakir** [comment]
@shaswat-indian Have you received a partner ID?

**Comment by shaswat-indian** [comment]
> @shaswat-indian Have you received a partner ID?

@ShahafBenYakir Not yet. We had submitted the technology partner form some time back but haven't received any update on that.

File: Packs/QueryAI/Integrations/QueryAI/QueryAI.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    if not query:
+        raise ValueError('Missing argument: "query"')
+
+    result = client.run_query(query=query, connection_params=connection_params, workflow_params=workflow_params,
+                              time_text=time_text, alias=alias)
+
+    drilldown_url = f'### Click here to [see details]({generate_drilldown_url(query, alias, time_text, workflow_params)})'
+    readable_output = tableToMarkdown(f'Query.AI Result for the query: {query}', result['data']) if result['data'] else ''
+    readable_output += drilldown_url
+    reply = {'result': result['data'] if result['data'] else result['reply'], 'markdown_string': readable_output}
```
Comment (guyfreund) [inline Packs/QueryAI/Integrations/QueryAI/QueryAI.py:134]: why are we adding the readable output to reply? which will be added to the context

**Review by guyfreund — CHANGES_REQUESTED** [review]
Overall looks very good, minor changes to be done here.

File: Packs/QueryAI/pack_metadata.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
+{
+    "name": "QueryAI",
+    "description": "Query.AI is a decentralized data access and analysis technology that simplifies security investigations across disparate platforms without data duplication.",
+    "support": "developer",
```
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Packs/QueryAI/pack_metadata.json:4]: But we haven't received any partner ID yet. Should we still change this?

File: Tests/conf.json [hunk 3]
```diff
@@ -2959,6 +2959,10 @@
             "playbookID": "FailedInstances - Test",
             "integrations": "Whois",
             "fromversion": "4.5.0"
+        },
+        {
+            "integrations": "Query.AI",
+            "playbookID": ""
```
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Tests/conf.json]: Reverted.

File: Packs/QueryAI/Integrations/QueryAI/QueryAI_description.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+Query.AI is a decentralized data access and analysis technology that simplifies security investigations across disparate platforms, without data duplication.
+
+In order to use this integration you need the following:
+1. The URL of Query.AI Proxy component (see below)
+2. An email registered with Query.AI belonging to your Organization
+3. The API key associated with above email
+4. Platform Connection Details of any platform integrated via Query.AI you wish to connect to (This can be overridden while executing commands)
+
+#### BASE_URL
+The base URL would be of the [Query.AI Proxy](https://proxy.query.ai:443) . Replace with hostname and port of the Query.AI Proxy component running in your environment.
```
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Packs/QueryAI/Integrations/QueryAI/QueryAI_description.md:11]: Added an example to make it more elaborate.

File: Packs/QueryAI/Integrations/QueryAI/QueryAI.yml [hunk 5]
```diff
@@ -0,0 +1,94 @@
+category: Analytics & SIEM
+commonfields:
+  id: Query.AI
+  version: -1
+configuration:
+- display: Query.AI Proxy URL
+  name: url
+  required: true
+  type: 0
+- display: Email ID
```
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Packs/QueryAI/Integrations/QueryAI/QueryAI.yml]: Done.

File: Packs/QueryAI/Integrations/QueryAI/QueryAI.yml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  required: true
+  type: 4
+- display: Default Platform Alias
+  name: alias
+  required: true
+  type: 0
+- display: Default Platform Connection Params. Eg- {"alias":{"username":"my_username","password":"my_password"}}
+  name: connection_params
+  required: true
+  type: 12
```
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Packs/QueryAI/Integrations/QueryAI/QueryAI.yml]: We've already given the example to make it clear how to enter the connection params.

File: Packs/QueryAI/pack_metadata.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Packs/QueryAI/pack_metadata.json:4]: This is your partner id: `query-ai-3526` you should've gotten an email about it, we'll check that.

File: Packs/QueryAI/Integrations/QueryAI/QueryAI.yml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+      isArray: false
+      name: connection_params
+      required: false
+      secret: false
+    - default: false
+      description: Workflow params as JSON object. Eg- {"param1":"value1","param2":"value2"}
+      isArray: false
+      name: workflow_params
+      required: false
+      secret: false
```
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Packs/QueryAI/Integrations/QueryAI/QueryAI.yml]: We've already given the example so that it's clear how to enter the connection params.

File: Packs/QueryAI/Integrations/QueryAI/QueryAI.yml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    description: Returns response for the query being run on Query.AI.
+    execution: false
+    name: queryai-run-query
+    outputs:
+    - contextPath: QueryAI.query.result
+      description: Response after running query
+      type: Unknown
+    - contextPath: QueryAI.query.markdown_string
+      description: Readable Response after running query
+      type: Unknown
```
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Packs/QueryAI/Integrations/QueryAI/QueryAI.yml]: Changed it to String.

File: Packs/QueryAI/Integrations/QueryAI/QueryAI.yml [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      required: false
+      secret: false
+    deprecated: false
+    description: Returns response for the query being run on Query.AI.
+    execution: false
+    name: queryai-run-query
+    outputs:
+    - contextPath: QueryAI.query.result
+      description: Response after running query
+      type: Unknown
```
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Packs/QueryAI/Integrations/QueryAI/QueryAI.yml:77]: This actually returns an Object/List which I don't see in the list of types allowed.

File: Packs/QueryAI/pack_metadata.json [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        "Threat Hunting",
+        "Incident Response",
+        "Analyst Assistive-AI"
+    ],
+    "useCases": [],
+    "keywords": [
+        "query",
+        "ai",
+        "query.ai"
+    ]
```
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Packs/QueryAI/pack_metadata.json]: Added.

File: Packs/QueryAI/Integrations/QueryAI/QueryAI.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
+import demistomock as demisto  # noqa: E402 lgtm [py/polluting-import]
+from CommonServerUserPython import *  # noqa: E402 lgtm [py/polluting-import]
+from CommonServerPython import *  # noqa: E402 lgtm [py/polluting-import]
```
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Packs/QueryAI/Integrations/QueryAI/QueryAI.py]: Removed comments.

File: Packs/QueryAI/Integrations/QueryAI/QueryAI.py (hunk 1 above)
Comment (shaswat-indian) [reply Packs/QueryAI/Integrations/QueryAI/QueryAI.py:134]: This is because for our use case we want to show that readable output in a field as well when called from a playbook, as it may contain some drill-down links to our product and the only way to set field values from a playbook is from the context data. Please see the attached screenshot in the description for reference.

File: Packs/QueryAI/Integrations/QueryAI/QueryAI.yml [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+      isArray: false
+      name: workflow_params
+      required: false
+      secret: false
+    - default: false
+      description: Search time period
+      isArray: false
+      name: time_text
+      required: false
+      secret: false
```
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Packs/QueryAI/Integrations/QueryAI/QueryAI.yml]: It can be in plain English. Something like "search 1 year ago to now". I've added this example in the arg description now.

File: Packs/QueryAI/pack_metadata.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (shaswat-indian) [reply Packs/QueryAI/pack_metadata.json]: Thanks, I've updated the PR description with it as well. Looking forward to the email.
